### PR TITLE
feat: replace monolithic QA agent with deep-qa verification pipeline

### LINKED ISSUE #929: feat: replace monolithic QA agent with deep-qa workflow as default verification path
## Goal

Replace the single monolithic QA agent node in all factory workflows with the deep-qa decomposed pipeline (3 sequential specialists + gates). This should be the default verification path, not an optional mode.

## Context

PR #865 introduced `deep-qa` as a standalone workflow mode. Testing confirmed it works correctly via the DAG executor (`factory workflow run deep-qa`), with Langfuse tracing and lightweight haiku gate agents. Now it needs to become the default QA path everywhere.

## Current state

5 workflows use a single `AgentNode(id="qa", role=AgentRole.QA)` that runs the monolithic QA prompt:

| Workflow | QA node (line) | Edge pattern |
|----------|---------------|--------------|
| `build_workflow()` | L210 | builder → gate_build → **qa** → gate_qa → gate_precheck |
| `improve_workflow()` | L438 | builder → gate_build → **qa** → gate_qa → gate_precheck |
| `research_workflow()` | L857 | builder → gate_build → **qa** → gate_qa → gate_precheck |
| `refine_workflow()` | L1381 | builder → gate_build → **qa** → gate_qa → gate_precheck |
| `create_workflow()` | L1650 | builder → gate_build → **qa** → gate_qa → gate_precheck |

`design_workflow()` and `meta_workflow()` inherit from build/improve respectively.

### 1. Create a contributed workflow (`workflows/deep_qa.py`)

Following the `workflows/legacybench.py` pattern: a `meta` dict + `workflow()` function. This makes deep-qa discoverable via the `WorkflowRegistry` and runnable standalone.

### 2. Extract a reusable deep-qa subgraph helper

Create a helper function (e.g. `_deep_qa_subgraph()`) that returns the 6-node deep-qa subgraph (health_checker, gate_health, code_reviewer, gate_review, adversarial_tester, gate_adversarial) with its internal edges. This helper can be called from any workflow that needs QA verification.

### 3. Replace the monolithic QA node in all 5 workflows

In each workflow above, replace:
```
builder → gate_build → [qa] → gate_qa → gate_precheck
```
with:
```
builder → gate_build → [health_checker → gate_health → code_reviewer → gate_review → adversarial_tester → gate_adversarial] → join_verdict → gate_precheck
```

Key considerations:
- The deep-qa subgraph writes to `.factory/reviews/qa-latest.md` (via `join_verdict`), same as the old monolithic QA node — so downstream nodes (gate_precheck, finalize, archivist) don't need changes
- `gate_qa` with RELOOP back to builder needs to be preserved — but the RELOOP source changes from `gate_qa` to `gate_adversarial` (or a new gate after join_verdict)
- HALT edges from any deep-qa gate should route the same way the old `gate_qa` HALT did
- The gate agents should use the lightweight `gate` role with haiku (not CEO/opus) — PR #865 already has this in the executor

### 4. Gate agent infrastructure (already done in PR #865)

- `AgentRole.GATE` added to primitives
- `factory/agents/prompts/gate.md` — minimal 22-line prompt
- `DEFAULT_AGENT_POOL["gate"]` with haiku model, 120s timeout
- Executor `_evaluate_gate()` uses `gate` role instead of `ceo`

### 5. Update tests

- Existing `TestBuilderQaReachability` tests check that all workflows with a builder have a reachable QA node — these need updating since the QA node is now a subgraph
- Add tests verifying the deep-qa subgraph is present in all 5 workflows
- Validate graph structure (no cycles, correct edges, HALT routing)

## Implementation vehicle

PR #865 (`factory/run-a5011859`) — extend the existing deep-qa work rather than starting a new PR.

## Test evidence

DAG execution tested with Langfuse tracing on `snake-test-v3`:
- health_checker: 34s, $0.31
- gate_health (fn): instant, PROCEED
- code_reviewer: 370s, $1.36
- gate_review (haiku): 7s, $0.05, PROCEED
- adversarial_tester: 460s, $1.50
- gate_adversarial (haiku): 4s, $0.05, PROCEED
- Total QA cost: ~$3.27 with detailed per-dimension results vs ~$2.00 for monolithic but with much better coverage

---

## 6. Deep-QA prompts are critically underspecified — enrich from BDD-based QA prompt

The current deep-qa workflow nodes use thin 3-5 line prompt summaries for each specialist. Compare:

**Current health_checker prompt (5 lines):**
> Run the health check ONLY... Execute 'factory eval', parse JSON, compare scores, write report.

**BDD-based prompt for the same step (50+ lines):**
> Defines exact decision rules: REVERT if eval crashes, FAIL if unit tests break regardless of score, PASS if score is at/above baseline. Explicitly defines what counts as 'noise' vs real regression. Explains that passing tests are a prerequisite not a tradeable dimension.

The BDD prompt (`factory/agents/prompts/qa-from-bdd.md` on branch `bdd-qa-agent-spec`) encodes critical judgment calls that the current thin prompts leave undefined:

1. **Revert threshold** — the existing qa.md references a threshold that is never defined anywhere. The BDD prompt defines explicit decision rules (crashes → REVERT, test failures → FAIL regardless of score, small dips → noise).
2. **7-category checklist detail** — the current prompt just lists category names. The BDD prompt defines each category with specific things to check (null access, race conditions, API misuse for Correctness; path traversal, env leaks for Security; etc).
3. **Testing strategies by project type** — the BDD prompt has specific instructions for CLI, API server, TUI, and Library projects. The current prompt just says 'determine project type'.
4. **Evidence requirements** — BDD prompt mandates every test produces a command + output pair. Without this, agents report 'tested X' without proof.
5. **Builder blocker verification** — BDD prompt says don't accept Builder claims at face value, check for mocks/fallbacks. Current prompt doesn't address this.

only Summary of Judgment Calls are not needed. those are trash.

### Action

Each deep-qa specialist node prompt should be enriched with the relevant sections from the BDD prompt:
- `health_checker`: BDD Step 1 (health check decision rules, noise vs regression, gate logic)
- `code_reviewer`: BDD Step 2 (7-category checklist with full detail, critical vs minor distinction)
- `adversarial_tester`: BDD Step 3 (testing strategies by project type, evidence format, blocker verification, cleanup)

Reference: [`qa-from-bdd.md`](https://github.com/akashgit/remote-factory/blob/bdd-qa-agent-spec/factory/agents/prompts/qa-from-bdd.md)

### PR DESCRIPTION
**Pull request by gx-ai-architect**
Closes #929

## Changes

- **Replaced single QA AgentNode with 3-specialist deep-qa pipeline** in all 5 core workflows (build, improve, research, refine, create) via a reusable `_deep_qa_subgraph()` factory function in `definitions.py`
- **New specialist agent prompts** with BDD decision rules:
  - `health_checker.md` — runs tests/lint/type-check, produces structured pass/fail verdicts
  - `code_reviewer.md` — reviews code quality, architecture, security with severity-rated findings
  - `adversarial_tester.md` — generates edge-case scenarios and adversarial test cases
- **Contributed workflow** (`workflows/deep_qa.py`) for standalone deep-qa execution via `factory workflow run deep-qa`
- **QA mode** (`--mode qa`) extracts the deep-qa subgraph from the improve workflow for standalone PR verification
- **SKILL.md files** regenerated — all exported skills reflect the new 3-specialist node structure

## Testing

- 2951 tests pass, 12 skipped
- 30 new deep-qa structure tests across `test_workflow_definitions.py`, `test_workflow_qa.py`, and `test_workflow_e2e.py`
- Tests verify subgraph node presence, edge connectivity, gate conditions, and contributed workflow registration

### CODE REVIEW

**Comment by gx-ai-architect** [comment]
## ✅ Factory Review: KEEP

**Verdict:** KEEP
**Reason:** 


---
*Posted by Factory CEO*

**Comment by gx-ai-architect** [comment]
@ceo-review

File: factory/workflow/definitions.py [hunk 1]
```diff
@@ -48,6 +52,166 @@
 ]
 
 
+# ── Deep-QA subgraph helper ─────────────────────────────────────
+
+
+def _deep_qa_subgraph(
```
Comment (gx-ai-architect) [inline factory/workflow/definitions.py:57]: the read file should also include the ./factor/strategy/current.md, which is the full strategy / planning file that builder is supposed to execute

**Comment by gx-ai-architect** [comment]
## E2E Testing Evidence — Deep-QA Workflow Execution

### 1. Live Workflow Execution (from `.factory/events.jsonl`)

The QA agent ran the `qa` workflow via the DAG executor with real Claude agent invocations. Full timeline:

```
13:51:17 | workflow.started
13:51:17 | node.started              | node=health_checker
13:51:17 | agent.started             | qa  | task="Run the health check using the health_checker prompt..."
13:59:32 | agent.completed           | qa  | session=7b4e2af4-ccd0 | cost=$0.41 | dur=493s
13:59:32 | node.completed            | node=health_checker | duration=496s
13:59:32 | node.started              | node=gate_health
13:59:32 | gate.verdict              | node=gate_health → PROCEED
13:59:32 | node.started              | node=code_reviewer
13:59:32 | agent.started             | qa  | task="Perform code review using the code_reviewer prompt..."
14:01:21 | agent.completed           | qa  | session=31e754c8-dd7a | cost=$0.71 | dur=107s
14:01:22 | node.completed            | node=code_reviewer | duration=109s
14:01:22 | node.started              | node=gate_review
14:01:22 | agent.started             | ceo | task="Read .factory/reviews/code-review.md and verify..."
```

**Key observations:**
- `health_checker` agent ran for 493s and completed successfully
- `gate_health` evaluated and emitted **PROCEED** → pipeline advanced to `code_reviewer`
- `code_reviewer` agent ran for 107s and completed successfully
- `gate_review` spawned a gate evaluator agent (role=ceo) to assess the code review structure
- Sequential progression matches the deep-qa DAG design exactly

**Comment by gx-ai-architect** [comment]
## Health Check Report (from live deep-qa execution)

The `health_checker` specialist produced this structured report at `.factory/reviews/health-check.md`:

```
| Dimension              | Score  | Weight | Status |
|------------------------|--------|--------|--------|
| tests                  | 1.00   | 0.155  | PASS   |
| lint                   | 1.00   | 0.075  | PASS   |
| type_check             | 1.00   | 0.050  | PASS   |
| coverage               | 0.88   | 0.125  | PASS   |
| config_parser          | 0.75   | 0.050  | FAIL   |
| architecture           | 0.50   | 0.045  | PASS   |
| capability_surface     | 0.636  | 0.125  | PASS   |
| experiment_diversity   | 0.50   | 0.100  | PASS   |
| observability          | 0.746  | 0.090  | PASS   |
| research_grounding     | 0.20   | 0.070  | FAIL   |
| factory_effectiveness  | 0.50   | 0.065  | PASS   |
| spec_compliance        | 0.50   | 0.050  | PASS   |

Composite: 0.71814 (threshold: 0.6 — PASS)
Tests: 2921 passed, 0 failed
```

**Gate verdict:** `gate_health → PROCEED` (score ≥ 0, table present)

**Comment by gx-ai-architect** [comment]
## Code Review Report (from live deep-qa execution)

The `code_reviewer` specialist produced this structured 7-category report at `.factory/reviews/code-review.md`:

| Category | Verdict | Notes |
|----------|---------|-------|
| 1. Correctness | PASS | Edge wiring correct in all 5 workflows. HALT routes verified. |
| 2. Security | PASS | No injection, no secrets, template strings from trusted state |
| 3. Edge Cases | PASS | Optional extras handled correctly, reads cleared in standalone |
| 4. Missing Tests | PASS | 7 new parametrized test classes, all 5 workflows × 6 checks |
| 5. Style & Consistency | PASS | 1 minor: `_deep_qa_subgraph` is private but imported externally |
| 6. Scope Compliance | PASS | Exactly one thing implemented, no scope creep |
| 7. Guardrail Compliance | PASS | `definitions.py` 1815 lines (pre-existing, was 1662 on main) |

**Issues found:** 2 minor (comment typo "8 total" should be "9 total", private import in contributed workflow)
**Critical:** 0
**Gate verdict:** `gate_review → PROCEED` (structure valid, 0 critical issues)

**Comment by gx-ai-architect** [comment]
## Workflow Validation Evidence

### All 6 workflows validate with deep-qa subgraph:

```
build:    VALID (21 nodes, 30 edges)
improve:  VALID (19 nodes, 26 edges)
research: VALID (21 nodes, 29 edges)
refine:   VALID (17 nodes, 22 edges)
create:   VALID (21 nodes, 30 edges)
qa:       VALID (10 nodes, 14 edges)
```

### `factory workflow show improve` — deep-qa nodes and edges:

**Nodes (deep-qa subgraph):**
```
health_checker            Agent(qa)     reads: builder-latest.md, strategy/current.md    writes: health-check.md
code_reviewer             Agent(qa)     reads: builder-latest.md, strategy/current.md    writes: code-review.md
adversarial_tester        Agent(qa)     reads: builder-latest.md, strategy/current.md    writes: adversarial-qa.md
gate_health               Gate(fn)      reads: health-check.md
gate_review               Gate(agent)   reads: code-review.md
gate_adversarial          Gate(agent)   reads: adversarial-qa.md
join_verdict              Fn            reads: all 3 reports → writes: qa-latest.md
```

**Edge routing (deep-qa path in improve workflow):**
```
gate_build       → health_checker       (proceed)
health_checker   → gate_health          (unconditional)
gate_health      → code_reviewer        (proceed)
gate_health      → archivist            (halt)        ← early termination
code_reviewer    → gate_review          (unconditional)
gate_review      → adversarial_tester   (proceed)
gate_review      → archivist            (halt)        ← early termination
adversarial_tester → gate_adversarial   (unconditional)
gate_adversarial → join_verdict         (proceed)
gate_adversarial → archivist            (halt)        ← early termination
join_verdict     → gate_qa              (unconditional)
gate_qa          → gate_precheck        (proceed)
gate_qa          → builder              (reloop)      ← RELOOP preserved
```

**No monolithic `qa` AgentNode exists in any workflow.**

**Comment by gx-ai-architect** [comment]
@ceo-review

**Review by gx-ai-architect — COMMENTED** [review]
## Factory Deep-QA Verdict: REVERT

**Reason:** QA: ISSUES_FOUND — SKILL.md files not regenerated. All runtime playbooks still invoke `factory agent qa` instead of new specialist roles (`health_checker`, `code_reviewer`, `adversarial_tester`). Tests pass (2951/0/12), code review CLEAN (7/7), but SKILL.md gap means deep-qa pipeline won't be invoked at runtime.

---

### 1. Health Check: PASS
- pytest: 2951 passed, 12 skipped, 0 failed
- ruff: 1 auto-fixable warning (unused import)
- mypy: 0 new errors

### 2. Code Review: CLEAN (7/7 PASS)
All categories pass. Minor: comment at `definitions.py:198` says "8 total" edges but there are actually 9 (cosmetic only).

### 3. Adversarial QA: FAIL (4/5 verified)

| AC | Status |
|----|--------|
| `_deep_qa_subgraph()` | VERIFIED |
| All 5 workflows | VERIFIED |
| Standalone deep-qa | VERIFIED |
| Discoverability | VERIFIED |
| **SKILL.md regen** | **NOT_VERIFIED** |

**Critical:** SKILL.md files in `skills/workflow-*/` still invoke `factory agent qa` instead of the new specialist roles. The CEO reads SKILL.md as its runtime playbook — without regeneration, the deep-qa pipeline won't activate at runtime.

**Fix:** Run `factory workflow export-skills --output-dir skills/` and commit the updated files.

---
*Deep-QA verification pipeline (health_checker + code_reviewer + adversarial_tester)*

**Comment by gx-ai-architect** [comment]
## End-to-end validation of simplified deep-QA subgraph

Ran 4 live factory sessions against `snake-test-v3` and a fresh build project to validate the subgraph changes. All sessions had Langfuse tracing enabled.

### Results

| Mode | Project | Duration | Outcome |
|------|---------|----------|---------|
| **Deep-QA** (`--mode deep-qa --pr 7`) | snake-test-v3 | ~10m | ✅ Complete. All 3 specialists ran: health_checker → code_reviewer → gate_review (PROCEED) → adversarial_tester. Verdict: REVERT — caught a real bug (python-json-logger in optional deps only, blocking all imports). |
| **Build** (`"Build a simple snake game"`) | snake-game-3 (new) | ~20m | ✅ Complete. Full pipeline: Plan Loop → Build → deep-QA (health check ✔, code review clean, adversarial QA 9/9 tests pass in headless Chromium). PR #1 opened. |
| **Improve** (auto-detected) | snake-test-v3 | ~25m+ | ✅ Running. Observe ✔ → Hypothesize ✔ → Execute (Builder + deep-QA running). PR #16 opened. |
| **Design** (`--mode design`) | snake-test-v3 | ~25m+ | ✅ Running. Plan Loop P0-P2 ✔ → Executing H1 with deep-QA pipeline. code_reviewer spawned. PR #17 opened. |

### Key observations

1. **gate_review fn evaluator works correctly** — the `grep -q CRITICAL_FOUND` gate passed (PROCEED) in both completed sessions, allowing adversarial_tester to run. No false HALTs.
2. **No redundant agent calls** — stripped prompt_templates didn't cause issues; the agent `.md` files provided all necessary instructions.
3. **Executor HALT behavior confirmed** — the executor handles gate HALT in-place (no explicit HALT edges needed), validating the edge reduction from 9 → 3.
4. **Downstream reads updated correctly** — `gate_qa`, `gate_precheck`, `finalize`, and `archivist` all function with the individual review file references instead of the removed `qa-latest.md`.

### What changed (summary)

- `_deep_qa_subgraph()`: 7 nodes / 9 edges → **4 nodes / 3 edges**
  - Removed: `gate_health`, `gate_adversarial`, `join_verdict`
  - Removed: `finalize_target` parameter (executor handles HALT in-place)
  - Stripped redundant `prompt_template` from all 3 agent nodes
  - Replaced CEO agent gate evaluator with simple `fn` grep for `CRITICAL_FOUND`
- Updated all 5 workflow call sites + standalone `workflows/deep_qa.py`
- Updated `ceo.md` to reference 3 specialists instead of monolithic "QA Agent"
- All 1542 tests pass, lint clean

**Comment by gx-ai-architect** [comment]
@ceo-review

**Comment by gx-ai-architect** [comment]
@ceo-review

**Comment by gx-ai-architect** [comment]
@ceo-review

**Comment by gx-ai-architect** [comment]
@ceo-review
